### PR TITLE
fix: Fix panics on GATs involving const generics

### LINKED ISSUE #11989: textDocument/semanticTokens/etc. GAT failure on latest pre-release
1. Create an empty binary cargo project with the nightly toolchain
2. Put the following code in main:

```
#![feature(generic_associated_types)]

pub trait Processor {
    type Ctx<'a, const N: usize>;

    fn process<const N: usize>(&mut self, ctx: Self::Ctx<'_, N>);
}

#[derive(Default, Debug)]
pub struct Phasor {
    phase: f32,
}

pub struct PhasorCtx<'a, const N: usize> {
    increments: &'a [f32; N],
    output: &'a mut [f32; N],
}

impl Processor for Phasor {
    type Ctx<'a, const N: usize> = PhasorCtx<'a, N>;

    fn process<const N: usize>(&mut self, ctx: Self::Ctx<'_, N>) {
        for i in 0..N {
            ctx.output[i] = self.phase;
            self.phase += ctx.increments[i];
            if self.phase > 1.0 {
                self.phase -= self.phase as i32 as f32;
            }
        }
    }
}

fn main() {
    println!("Hello, world!");
}
```

I'm getting errors like `textDocument/semanticTokens/full`, `textDocument/semanticTokens/full/delta`, `textDocument/inlayHints` and `textDocument/definition` failed every time I start typing.

I know several GAT issues have already been solved, but I'm working with a pre-release version from 2 days ago. Assuming [this fix](https://github.com/rust-lang/rust-analyzer/pull/11957) already made it in 3 days ago I'm still running into issues and thought this warranted an issue report. :)

**rust-analyzer version**: rust-analyzer version: faaec86ff 2022-04-12 nightly
**rustc version**: rustc 1.61.0-nightly (c84f39e6c 2022-03-20)

### LINKED ISSUE #12193: GAT + const generics crash v2
unfortunately, I'm hitting another of those… (previously: #11769)
```rust
#![feature(generic_associated_types)]

trait Collection {
    type Item;
    type Member<T>: Collection<Item = T>;
    fn add(&mut self, value: Self::Item) -> Result<(), Self::Error>;
}


struct ConstGen<T, const N: usize> {
    data: [T; N],
}

impl<T, const N: usize> Collection for ConstGen<T, N> {
    type Item = T;
    type Member<U> = ConstGen<U, N>;
}
```

**rust-analyzer version**: `5d5bbec9b 2022-05-08 nightly`

**rustc version**: `rustc 1.62.0-nightly (e1b71feb5 2022-05-03)`

**relevant settings**: `None`

crash log:

```code
Panic context:
> 
version: 5d5bbec9b 2022-05-08 nightly
request: textDocument/inlayHint InlayHintParams {
    work_done_progress_params: WorkDoneProgressParams {
        work_done_token: None,
    },
    text_document: TextDocumentIdentifier {
        uri: Url {
            scheme: "file",
            cannot_be_a_base: false,
            username: "",
            password: None,
            host: None,
            port: None,
            path: "/Users/ace/play/gat-coll/src/lib.rs",
            query: None,
            fragment: None,
        },
    },
    range: Range {
        start: Position {
            line: 0,
            character: 0,
        },
        end: Position {
            line: 17,
            character: 0,
        },
    },
}

thread '<unnamed>' panicked at 'mismatched kinds in substitution', /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/chalk-ir-0.82.0/src/fold/subst.rs:64:22
stack backtrace:
   0: std::panicking::begin_panic
   1: <chalk_ir::fold::subst::Subst<I> as chalk_ir::fold::Folder<I>>::fold_free_var_ty
   2: <chalk_ir::Ty<I> as chalk_ir::fold::SuperFold<I>>::super_fold_with
   3: chalk_ir::fold::boring_impls::<impl chalk_ir::fold::Fold<I> for chalk_ir::GenericArg<I>>::fold_with
   4: <smallvec::SmallVec<A> as core::iter::traits::collect::Extend<<A as smallvec::Array>::Item>>::extend
   5: chalk_ir::fold::boring_impls::<impl chalk_ir::fold::Fold<I> for chalk_ir::Substitution<I>>::fold_with
   6: <chalk_ir::Ty<I> as chalk_ir::fold::SuperFold<I>>::super_fold_with
   7: chalk_ir::fold::boring_impls::<impl chalk_ir::fold::Fold<I> for chalk_ir::GenericArg<I>>::fold_with
   8: <smallvec::SmallVec<A> as core::iter::traits::collect::Extend<<A as smallvec::Array>::Item>>::extend
   9: chalk_ir::fold::boring_impls::<impl chalk_ir::fold::Fold<I> for chalk_ir::Substitution<I>>::fold_with
  10: chalk_ir::fold::subst::Subst<I>::apply
  11: chalk_ir::Binders<T>::substitute
  12: chalk_solve::split::Split::impl_parameters_and_projection_from_associated_ty_value
  13: chalk_solve::clauses::builder::ClauseBuilder<I>::push_binders
  14: chalk_solve::clauses::push_program_clauses_for_associated_type_values_in_impls_of
  15: chalk_recursive::solve::SolveIterationHelpers::solve_from_clauses
  16: chalk_recursive::solve::SolveIteration::solve_iteration
  17: chalk_recursive::fixed_point::RecursiveContext<K,V>::solve_goal
  18: chalk_recursive::fulfill::Fulfill<I,Solver>::prove
  19: chalk_recursive::fulfill::Fulfill<I,Solver>::solve
  20: chalk_recursive::solve::SolveIterationHelpers::solve_from_clauses
  21: chalk_recursive::solve::SolveIteration::solve_iteration
  22: chalk_recursive::fixed_point::RecursiveContext<K,V>::solve_goal
  23: chalk_recursive::fixed_point::RecursiveContext<K,V>::solve_root_goal
  24: <hir_ty::db::TraitSolveQueryQuery as salsa::plumbing::QueryFunction>::execute
  25: salsa::runtime::Runtime::execute_query_implementation
  26: salsa::derived::slot::Slot<Q,MP>::read_upgrade
  27: salsa::derived::slot::Slot<Q,MP>::read
  28: salsa::QueryTable<Q>::get
  29: <DB as hir_ty::db::HirDatabase>::trait_solve_query
  30: hir_ty::db::trait_solve_wait
  31: <DB as hir_ty::db::HirDatabase>::trait_solve
  32: hir_ty::infer::unify::InferenceTable::register_obligation_in_env
  33: hir_ty::infer::unify::InferenceTable::normalize_projection_ty
  34: <hir_ty::fold_tys_and_consts::TyFolder<F> as chalk_ir::fold::Folder<hir_ty::interner::Interner>>::fold_ty
  35: hir_ty::infer::infer_query
  36: salsa::runtime::Runtime::execute_query_implementation
  37: salsa::derived::slot::Slot<Q,MP>::read_upgrade
  38: salsa::derived::slot::Slot<Q,MP>::read
  39: salsa::QueryTable<Q>::get
  40: <DB as hir_ty::db::HirDatabase>::infer_query
  41: hir_ty::db::infer_wait
  42: hir::source_analyzer::SourceAnalyzer::new_for_body
  43: hir::semantics::SemanticsImpl::analyze_impl
  44: hir::semantics::SemanticsImpl::type_of_pat
  45: ide::inlay_hints::hints
  46: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
  47: ide::inlay_hints::inlay_hints
  48: std::panicking::try
  49: rust_analyzer::handlers::handle_inlay_hints
  50: std::panicking::try
  51: <F as threadpool::FnBox>::call_box
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[Error - 3:38:50 AM] Request textDocument/inlayHint failed.
  Message: request handler panicked: mismatched kinds in substitution
  Code: -32603 

```

### PR DESCRIPTION
**Pull request by N3xed**
This workaround avoids constant crashing of rust analyzer when using GATs with const generics,
even when the const generics are only on the `impl` block.

The workaround treats GATs as non-existing if either itself or the parent has const generics and
removes relevant panicking code-paths.

Fixes #11989, fixes #12193

### CODE REVIEW

File: crates/hir-ty/src/utils.rs [hunk 1]
```diff
@@ -264,12 +279,11 @@ impl Generics {
...
-            let (idx, (_local_id, data)) = self
-                .params
-                .iter()
-                .enumerate()
-                .find(|(_, (idx, _))| *idx == param.local_id)
-                .unwrap();
+            let (idx, (_local_id, data)) =
+                match self.params.iter().enumerate().find(|(_, (idx, _))| *idx == param.local_id) {
+                    None if self.is_gat => return None,
+                    val => val.unwrap(),
```
Comment (flodiebold) [inline crates/hir-ty/src/utils.rs:285]: I don't know why we have `find_param` return an option if we never actually return `None`. Let's just change it to return `None` if the param isn't found in all cases, and get rid of `is_gat`?

File: crates/hir-ty/src/lower.rs [hunk 2]
```diff
@@ -509,7 +509,14 @@ impl<'a> TyLoweringContext<'a> {
                         TyKind::Placeholder(to_placeholder_idx(self.db, param_id.into()))
                     }
                     ParamLoweringMode::Variable => {
-                        let idx = generics.param_idx(param_id.into()).expect("matching generics");
+                        let idx = match generics.param_idx(param_id.into()) {
+                            // FIXME: never crash if this involves const generic associated types
+                            None if generics.is_gat => {
```
Comment (flodiebold) [inline crates/hir-ty/src/lower.rs:514]: let's just do a `never!()` check here instead of checking `is_gat`

Comment (N3xed) [reply crates/hir-ty/src/lower.rs:514]: This will make the test fail though.

Comment (flodiebold) [reply crates/hir-ty/src/lower.rs:514]: Yeah, I would be fine with that. It's a bit unfortunate that we can't test that it doesn't crash, but I'd rather have the crashing test as a reminder that it's still broken.

File: crates/hir-ty/src/consteval.rs [hunk 3]
```diff
@@ -364,6 +364,8 @@ pub(crate) fn path_to_const(
                 }
                 ParamLoweringMode::Variable => match args.param_idx(p.into()) {
                     Some(x) => ConstValue::BoundVar(BoundVar::new(debruijn, x)),
+                    // FIXME: never crash if this involves const generic associated types
```
[the remark this answers is not in the document]
Comment (N3xed) [reply crates/hir-ty/src/consteval.rs]: Done.

**Comment by flodiebold** [comment]
LGTM, but the test of course fails :sweat_smile: 

@bors delegate+

**Comment by N3xed** [comment]
Hopefully, the CI failing isn't a big issue. Unfortunately, I'm not familiar enough to tackle the underlying problem since I'm very new to the rust analyzer code base, and just implementing this workaround took me quite a while.

Nonetheless thanks for the review.
@bors r+

File: crates/hir-ty/src/tests/regression.rs [hunk 4]
```diff
@@ -1526,6 +1526,34 @@ unsafe impl Storage for InlineStorage {
...
 }
 
+#[test]
+fn gat_crash_3() {
+    // FIXME: This test currently crashes rust analyzer in a debug build but not in a
+    // release build (i.e. for the user). With the assumption that tests will always be run
+    // in debug mode, we catch the unwind and expect that it panicked. See the
+    // [`crate::utils::generics`] function for more information.
+    cov_mark::check!(ignore_gats);
+    std::panic::catch_unwind(|| {
```
Comment (N3xed) [inline crates/hir-ty/src/tests/regression.rs:1536]: @flodiebold So, to make the test succeed I just catch the unwind here right now, which is pretty much the same as `#[shoud_panic]`. Or can we merge without passing tests?

**Pull request merged**